### DISCUSSION TITLE
[Proposal] Allow unboxing of Nullable<T> with HasValue and/or "is not null" pattern

### DISCUSSION THREAD

**Discussion by datvm** [dotnet/csharplang#4645] [Language Ideas]
# Motivation

Currently after checking if a `Nullable<T>` has value or not (is null or not), we need another formal line to declare a non-nullable value of it:

```
if (skipId.HasValue)
{
	var v = skipId.Value;
	query = query
		.Where(q => q.Id > v);

	// Other code that use v
}
```

# Proposal

Allow declaration of a variable right after `Nullable<T>.HasValue` property and/or in a `is not null` pattern:

```
if (skipId.HasValue v)
{
	query = query
		.Where(q => q.Id > v);

	// Other code that use v
}

if (skipId is not null v)
{
	query = query
		.Where(q => q.Id > v);

	// Other code that use v
}
```

**Comment by jnm2** [comment]
The syntax that works today (C# 8+) is `if (skipId is { } v)` but it's not as intuitive.

@CyrusNajmabadi You talked about being interested in `is not null v` a while back. What do you think?

**Reply by datvm** [reply]
Just want to chime in that I am more favored of `skipId.HasValue v` because it also feels more natural (skipId has value v) but I am okay with either or both.

**Reply by CyrusNajmabadi** [reply]
I'm still interested in it. But it's way way down on my list of what I think is important
